### DISCUSSION TITLE
Speckles in the canvas

### DISCUSSION THREAD

**Discussion by gktval** [mono/SkiaSharp#1618] [Q&A]
I have an SKContol in a Windows Form. After drawing the bitmap to the canvas and Invalidating the control, I am often seeing speckles in the image similar to this:

![image](https://user-images.githubusercontent.com/7605046/54769076-1fc32000-4bc6-11e9-9b65-ed846b2de812.png)

This happens on the PaintSurfaceEvent. I also see these random speckles when drawing an empty bitmap to the canvas. Here is the code I use to draw to the bitmap to the canvas in the paint event:

```
Dim canvas As SKCanvas = e.Surface.Canvas
canvas.Clear(SKColors.White)
Dim bitmap As New SKBitmap(e.Info.Rect.Width, e.Info.Rect.Height)
canvas.DrawBitmap(bitmap, 0, 0)
```

I can also verify that the speckles also appear on Android device too, not just windows.
Is there a work around to get rid of this, or is this a problem with drawing?

**Comment by mattleibow** [comment]
This is because both the surface canvas an the bitmap needs to be cleared. In you code, you are clearing the canvas, and then drawing a uninitialized bitmap.

This may just be because this is just a snippet, but what is the reason for creating the bitmap?

If the drawing in the bitmap is not transparent, such as with clear white, then you don't need to clear the surface - saving on that processing.

**Comment by gktval** [comment]
Thanks for the reply. The reason for creating the bitmap is so that I can draw multiple images and vectors to it. Think of a map, where the user loads and elevation layer, then an aerial image, the some vector layers such as a boundary or a road. On top of that, another bitmap is drawn with the gps cursor (see example above in my previous post) which gets refreshed every second depending on how far the user has moved.

In my actual code, I initialize the bitmap separately and then draw all the map layers to it, Then I draw the final bitmap to the canvas. I apologize for not showing the entire code above, but I can reproduce the same results with the code below:

    `Protected Overrides Sub OnPaintSurface(e As SKPaintSurfaceEventArgs)
        Dim clip As New Rectangle(e.Info.Rect.Location.X, e.Info.Rect.Location.Y,
                                  e.Info.Rect.Size.Width, e.Info.Rect.Size.Height)
        If clip.Width < 1 OrElse clip.Height < 1 Then Return 'Nothing to paint

        Dim bitmap As New SKBitmap(clip.Width, clip.Height)
        Dim canvas As SKCanvas = e.Surface.Canvas
        canvas.Clear(SKColors.White)
        canvas.DrawBitmap(bitmap, 0, 0)

        MyBase.OnPaintSurface(e)
    End Sub`

I did a bitmap.Erase(SkColors.White) and it does seem to help considerably. But this also seems a bit redundant to have to do this on a newly created bitmap.

After further testing, the issue is better with bitmap.Erase method, but I still see some subtle speckles near the top of the canvas
![image](https://user-images.githubusercontent.com/7605046/58269960-de75fa80-7d45-11e9-92c3-8572236a23c1.png)

**Comment by claunia** [comment]
Hi, I confirm this issue as explained in https://github.com/wieslawsoltes/Svg.Skia/issues/28

While it is expected a newly initialized element in C contains random memory you expect the contrary in C#, SKCanvas should be clear on `new`.

Regards

**Comment by artemiusgreat** [comment]
You might try [clearing both](https://github.com/Indemos/Canvas-V1/blob/master/Controls/CanvasImageControl.cs#L178), bitmap and canvas, before drawing. 

```C#
skBitmap.Erase(SKColors.Transparent);
skCanvas.Clear(SKColors.Transparent);
```
